src/core/agent/services/keriaNotificationService.ts

```diff
@@ -0,0 +1,743 @@
+        const status = error.message.split(" - ")[1];
+        if (/404/gi.test(status)) {
+          return [];
+        } else {
+          throw error;
+        }
+      });
+    if (!multisigNotification || !multisigNotification.length) {
+      await this.markNotification(notif.i);
+      return false;
+    }
+    const multisigId = multisigNotification[0]?.exn?.a?.gid;
+    if (!multisigId) {
+      await this.markNotification(notif.i);
+      return false;
+    }
+    const hasMultisig = await this.multiSigs.hasMultisig(multisigId);
+    const notificationsForThisMultisig =
+      await this.findNotificationsByMultisigId(multisigId);
+    if (hasMultisig || notificationsForThisMultisig.length) {
+      await this.markNotification(notif.i);
+      return false;
+    }
+    return true;
+  }
+
+  private async processMultiSigExnNotification(
+    notif: Notification
+  ): Promise<boolean> {
+    const exchange = await this.props.signifyClient.exchanges().get(notif.a.d);
+
+    if (exchange?.exn?.e?.exn?.r !== ExchangeRoute.IpexAdmit) {
+      await this.markNotification(notif.i);
+      return false;
+    }
+
+    const existMultisig = await this.identifiers.getIdentifier(
+      exchange?.exn?.e?.exn?.i
+    );
+    if (!existMultisig) {
+      await this.markNotification(notif.i);
+      return false;
+    }
+
+    const previousExnGrantMsg = await this.props.signifyClient
+      .exchanges()
+      .get(exchange?.exn.e.exn.p);
+
+    const existingCredential = await this.props.signifyClient
+      .credentials()
+      .get(previousExnGrantMsg.exn.e.acdc.d)
+      .catch(() => undefined);
+
+    if (existingCredential) {
+      await this.markNotification(notif.i);
+      return false;
+    }
+    return true;
+  }
+
+  private async processExnIpexAgreeNotification(
+    notif: Notification
+  ): Promise<boolean> {
+    const existingLinkedIpexRecord = await this.ipexMessageStorage
+      .getIpexMessageMetadata(notif.a.d)
+      .catch((error) => {
+        if (
+          error.message ===
+          IpexMessageStorage.IPEX_MESSAGE_METADATA_RECORD_MISSING
+        ) {
+          return undefined;
+        } else {
+          throw error;
+        }
+      });
+    if (!existingLinkedIpexRecord) {
+      const exchange = await this.props.signifyClient
+        .exchanges()
+        .get(notif.a.d);
+      await this.ipexCommunications.createLinkedIpexMessageRecord(
+        exchange,
+        ConnectionHistoryType.CREDENTIAL_REQUEST_AGREE
+      );
+    }
+    await this.ipexCommunications.grantAcdcFromAgree(notif.a.d);
+    await this.markNotification(notif.i);
+    return false;
+  }
+
+  private async createNotificationRecord(
+    event: Notification
+  ): Promise<KeriaNotification> {
+    const exchange = await this.props.signifyClient.exchanges().get(event.a.d);
+
+    const metadata: any = {
+      id: event.i,
+      a: event.a,
+      read: false,
+      route: event.a.r,
+      connectionId: exchange.exn.i,
+    };
+    if (
```

Comment:
spacing in this function is a bit tight, please add whitespace around this if statement for readability